### PR TITLE
feat(core): make GET /api/swagger.json contain all api routes

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

Make GET /api/swagger.json contain all `/api/*` routes

( Next PR: LOG-2809 Convert the Zod types (`ZodUnion`, `ZodUnknown` and etc.) to OpenAPI schemas )

## Linear Issue Reference

LOG-2609
LOG-2811

## Testing

image

image

image

TODO:

- [x] missing some routes
- [x] test swaggerRoutes
- [x] test zodTypeToSwagger
- [x] refactor `GET /api/swagger.json` (easy to understand)

### CODE REVIEW

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [inline packages/core/src/routes/swagger.ts:71]: This doesn't feel right.  As the original route schema flatten the methods internally, why do we still have the cases you listed above? Should be handled at the root route level

**Review by simeng-li — CHANGES_REQUESTED** [review]

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:12]: The first `router` is used to register `GET /api/swagger.json`,
and the `otherRouters` are required to generate swagger.json containing all routes.

Comment (IceHe) [reply packages/core/src/routes/swagger.ts]: Refactored. Replace `otherRouters` with `allRouters`. PTAL.

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [inline packages/core/src/routes/swagger.ts:66]: Set a map outside and use for .. in ... to loop through the routes? Avoid disabling this lint rule.

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [inline packages/core/src/routes/swagger.ts:80]: nit: what's the difference between {} and Object.create(null)?

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [inline packages/core/src/routes/swagger.ts:75]: make your pathObject (need a better name, pathObject is confusing) as a map. Then :

```js
pathsObject.set([path], { ... pathsObject.get(path), [method]: operation });
```

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:12]: There is no reusable type in koa-router.

A. We can infer the koa-router supported methods via `keyof Router`, but they're mixed with some unnecessary keys.
B. We can reuse the dependency `methods`, but it contains many methods that koa-router does not support.

So I decided to declare it according to koa-router implementation by ourselves.

image

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:80]: image

`Object.create()` will not be used after refactoring according to other comments

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:75]: Refactored. PTAL

```ts
    const pathMap = new Map<string, MethodMap>();

    // Group routes by path
    for (const { path, method, operation } of routes) {
      pathMap.set(path, { ...pathMap.get(path), [method]: operation });
    }

    const document: OpenAPIV3.Document = {
      ……,
      paths: Object.fromEntries(pathMap),
    };
```

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:66]: Refactored. PTAL

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:18]: Refactored. PTAL.

```ts
type MethodMap = {
  [key in HttpMethod]?: OpenAPIV3.OperationObject;
};
```

and `PathMap` is no longer needed now.

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:13]: Refactored. Replaced `HttpMethod` with `OpenAPIV3.HttpMethods`.

```ts
    enum HttpMethods {
        GET = "get",
        PUT = "put",
        POST = "post",
        DELETE = "delete",
        OPTIONS = "options",
        HEAD = "head",
        PATCH = "patch",
        TRACE = "trace"
    }
```

Comment (simeng-li) [reply packages/core/src/routes/swagger.ts:10]: try
`
Pick<Router.Layer, 'path' | 'methods'> & { operation: OpenAPIV3.OperationObject; }
`

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [reply packages/core/src/routes/swagger.ts]: I mean under your case, not following why have to be `Object.create(null)`

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [reply packages/core/src/routes/swagger.ts]: Please make sure to test before merging.  I just posted a proposal that hasn't been tested or verified.

File: packages/core/src/routes/swagger.ts
Comment (simeng-li) [reply packages/core/src/routes/swagger.ts]: To be more clear. Whenever a js object is initially assigned it contains the prototype properties by default, is there any specific reasons you want to remove it.

**Review by simeng-li — APPROVED** [review]
unblock you.  Overall looks good. Please make sure to test well before merging and address the comments.  
P.S. still same suggestions as before, thinks twice and design well before actually writing down any code.  Carefully reviewing the existing code or imported lib ahead, we should make logic more simple and straight forward not more complicated.

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts]: i create it with no properties from Object.prototype on purpose, so we can use it like a simple map/dict. under this case, it's enough to use.

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts:10]: > try `Pick<Router.Layer, 'path' | 'methods'> & { operation: OpenAPIV3.OperationObject; }`

It does not work

File: packages/core/src/routes/swagger.ts
Comment (IceHe) [reply packages/core/src/routes/swagger.ts]: It works. Thx.

**Pull request merged**
